metrics/influx.go

```diff
@@ -0,0 +1,165 @@
+}
+
+type ConnectionMetrics struct {
+	ActiveConnections      metrics.Counter
+	ConnectionCloseErrors  metrics.Meter
+	ConnectionAcceptErrors metrics.Meter
+}
+
+func NewInfluxMetricsEntry(name string, r metrics.Registry) *MetricsEntry {
+	return &MetricsEntry{
+		Duration:   metrics.GetOrRegisterTimer(fmt.Sprintf("%s.request_duration", name), r),
+		Errors:     metrics.GetOrRegisterMeter(fmt.Sprintf("%s.error_count", name), r),
+		BadRequest: metrics.GetOrRegisterMeter(fmt.Sprintf("%s.bad_request_count", name), r),
+		Request:    metrics.GetOrRegisterMeter(fmt.Sprintf("%s.request_count", name), r),
+	}
+}
+
+func NewInfluxMetricsEntryBackendPuts(name string, r metrics.Registry) *MetricsEntryByFormat {
+	return &MetricsEntryByFormat{
+		Duration:       metrics.GetOrRegisterTimer(fmt.Sprintf("%s.request_duration", name), r),
+		Errors:         metrics.GetOrRegisterMeter(fmt.Sprintf("%s.error_count", name), r),
+		BadRequest:     metrics.GetOrRegisterMeter(fmt.Sprintf("%s.bad_request_count", name), r),
+		JsonRequest:    metrics.GetOrRegisterMeter(fmt.Sprintf("%s.json_request_count", name), r),
+		XmlRequest:     metrics.GetOrRegisterMeter(fmt.Sprintf("%s.xml_request_count", name), r),
+		DefinesTTL:     metrics.GetOrRegisterMeter(fmt.Sprintf("%s.defines_ttl", name), r),
+		InvalidRequest: metrics.GetOrRegisterMeter(fmt.Sprintf("%s.unknown_request_count", name), r),
+		RequestLength:  metrics.GetOrRegisterHistogram(name+".request_size_bytes", r, metrics.NewExpDecaySample(1028, 0.015)),
+	}
+}
+
+func NewInfluxConnectionMetrics(r metrics.Registry) *ConnectionMetrics {
+	return &ConnectionMetrics{
+		ActiveConnections:      metrics.GetOrRegisterCounter("connections.active_incoming", r),
+		ConnectionAcceptErrors: metrics.GetOrRegisterMeter("connections.accept_errors", r),
+		ConnectionCloseErrors:  metrics.GetOrRegisterMeter("connections.close_errors", r),
+	}
+}
+
+type InfluxMetrics struct {
+	Registry        metrics.Registry
+	Puts            *MetricsEntry
+	Gets            *MetricsEntry
+	PutsBackend     *MetricsEntryByFormat
+	GetsBackend     *MetricsEntry
+	Connections     *ConnectionMetrics
+	ExtraTTLSeconds metrics.Histogram
+}
+
+// Export begins sending metrics to the configured database.
+// This method blocks indefinitely, so it should probably be run in a goroutine.
+func (m InfluxMetrics) Export(cfg config.Metrics) {
+	logrus.Infof("Metrics will be exported to Influx with host=%s, db=%s, username=%s", cfg.Influx.Host, cfg.Influx.Database, cfg.Influx.Username)
+	influxdb.InfluxDB(
+		m.Registry,          // metrics registry
+		time.Second*10,      // interval
+		cfg.Influx.Host,     // the InfluxDB url
+		cfg.Influx.Database, // your InfluxDB database
+		cfg.Influx.Username, // your InfluxDB user
+		cfg.Influx.Password, // your InfluxDB password
+	)
+	return
+}
+
+func (m InfluxMetrics) Increment(metricName string, start *time.Time, value string) {
+	switch metricName {
+	case "puts.current_url.request_duration":
```

Comment:
Why do you prefer to use string identifiers here? Consider breaking these out into methods and let the metric provider use their discretion. I think that approach works well in Prebid Server, with exception of the very large interface contract.

This gives more freedom for Prometheus to record things differently from Influx under the covers.

Reply:
Done